### PR TITLE
(GH-2738) Developer update about script loading changes

### PR DESCRIPTION
**Pull request by lucywyman**
This includes a developer update detailing the motivation for changing
the syntax for loading scripts and files, and the phased approach we're
taking to rolling changes out.

!feature

* **Developer Update: Script loading changes**

  There's a new Developer Update in town, [read it
here](https://puppet.com/docs/bolt/latest/developer_updates.html).

### CODE REVIEW

**Comment by hestonhoffman** [comment]
@lucywyman I put a PR up to your branch here: https://github.com/lucywyman/bolt/pull/12

**Comment by lucywyman** [comment]
Merged! I did see an interesting way to refer to the value passed in by users, which is as a "file pattern". I like that better than like "module syntax", but I'm not sure if that will just be more confusing for people since it's not a common phrase.

**Comment by hestonhoffman** [comment]
I'm not sure about calling out "specific module-style syntax". I'm trying to think of a different way to phrase this, but I'm also wondering if the change we're making needs a name?

File: documentation/developer_updates.md [hunk 1]
```diff
@@ -3,6 +3,111 @@
...
+All of these loading methods leave out one of the most common Bolt artifacts:
+scripts. You can load a script using module-style file loading, but the script
+needs to be in the `files/` subdirectory of a module, not the `scripts/`
+subdirectory. It's about time scripts got their due as the Bolt workhorses they
+are. That's why we're introducing specific module-style syntax for scripts.
+While "specific module-style syntax" is kind of a mouthful, it will provide a
+more elegant and comprehensible workflow for sharing and loading scripts from a
+module. With these changes, there's no magic to understanding where Bolt is
+loading a script from, and scripts are more visible as the core Bolt entities
+that they are.
```
Comment (hestonhoffman) [inline documentation/developer_updates.md:40]:
```suggestion
are. That's why we're introducing a more elegant and comprehensible workflow for sharing and loading scripts from a module. With these changes, there's no magic to understanding where Bolt is loading a script from, and scripts are more visible as the core Bolt entities
that they are.
```

File: documentation/developer_updates.md [hunk 2]
```diff
@@ -3,6 +3,111 @@
...
+
+Eventually, users will be able to load files by providing
+`<module_name>/files/myfile`, and load scripts by providing
+`<module_name>/scripts/myscript.sh`. This more detailed syntax will make scripts
+first-class citizens in Bolt, and disambiguate loading files vs loading scripts
+from a module. Plus, it is easier to understand where files are coming from if
+the full path is provided, as opposed to magically looking in the `files/`
+directory for a file. Once this feature is fully rolled out, the old, magic
+syntax `<module_name>/myfile` will no longer load from the `files/` subdirectory
+of a module.
```
Comment (hestonhoffman) [inline documentation/developer_updates.md:52]:
```suggestion
Eventually, you'll be able to load scripts and files using their fully qualified
names. In other words, you'll provide Bolt with a more detailed path to 
the script or file within a module. Files will remain in a module's `files` 
subdirectory, but scripts will live in a new `scripts` subdirectory. For 
example, to load a file you'll provide Bolt with a 
`<module_name>/files/myfile` path, and to load a script, you'll use the
syntax `<module_name>/scripts/myscript.sh`.

This more detailed syntax will make scripts
first-class citizens in Bolt, and disambiguate loading files vs loading scripts
from a module. Plus, it is easier to understand where files are coming from if
the fully qualified name is provided, as opposed to magically looking in the `files/`
directory for a file. Once this feature is fully rolled out, the old, magic
syntax `<module_name>/myfile` will no longer load from the `files/` subdirectory
of a module.
```

File: documentation/developer_updates.md [hunk 3]
```diff
@@ -3,6 +3,111 @@
...
+#### How we get there 🚲
+
+This is a big change in how Bolt loads file and scripts, so it's going to take a few different
+phases to transition from where we are to where we're going. Inevitably, in order for a module to
+function with older versions of Bolt, module authors will need to have scripts in two places at some
+point: one under the `files/` directory and one in the new location that Bolt will load from. We
+know this is a burden on content authors, but we think it's worth the improved experience for users
+new to the Puppet ecosystem and Bolt.
+
+**Phase 1**: In this phase, specific module-style syntax will be opt-in using the new `file_paths`
```
Comment (hestonhoffman) [inline documentation/developer_updates.md:63]:
```suggestion
**Phase 1**: In this phase, fully qualified names will be opt-in using the new `file_paths`
```

File: documentation/developer_updates.md [hunk 4]
```diff
@@ -3,6 +3,111 @@
...
+- Provided: `module/scripts/myscript.sh`
+  - Load from `scripts/myscript.sh` (new behavior)
+  - Fall back to `files/scripts/myscript.sh` (old behavior)
+- Provided: `module/files/myscript.sh`
+  - Load from `files/myscript.sh` (new behavior)
+  - Fall back to `files/files/myscript.sh` (old behavior)
+
+**Phase 3**: In this phase, we will issue a deprecation warning when Bolt finds
+a script in a `files` directory, or if you provide Bolt with a non-specific
+module-style path.
```
Comment (hestonhoffman) [inline documentation/developer_updates.md:92]:
```suggestion
a script in a `files` directory, or if you provide Bolt with the old
module-style path.
```

File: documentation/developer_updates.md [hunk 5]
```diff
@@ -3,6 +3,111 @@
...
+`<module_name>` is the name of the module on the modulepath to load from, and
+`<path_to_file>` is the path to the file within the `<module_name>/files/`
+subdirectory. For example, if you provided Bolt with the path
+`ruby_task_helper/task_helper.rb`, Bolt would find the `ruby_task_helper` module
+on the modulepath, and look in `ruby_task_helper/files/` for the file
+`task_helper.rb`. Similarly, provided the path
+`powershell_task_helper/BoltPwshHelper/BoltPwshHelper.psm1`, Bolt looks in
+`powershell_task_helper/files/BoltPwshHelper/` for `BoltPwshHelper.psm1`.
+
+Users can also load tasks and plans as first-class citizens of a module, but with a slightly
```
Comment (hestonhoffman) [inline documentation/developer_updates.md:25]:
```suggestion
You can also load tasks and plans as first-class citizens of a module, but with a slightly
```

**Comment by lucywyman** [comment]
Sweet! Added those edits, and also read through and tried to make our wording more consistent plus added a short graf about our timeline (or lack thereof).

**Review by hestonhoffman — APPROVED** [review]
Looks good! :)

**Pull request merged**